File: frontend/lib/action-center-exit.ts
import {
  buildActionCenterPermissionEnvelope,
  buildActionCenterWorkspaceSummary,
  type ActionCenterAssignment,
  type ActionCenterDossier,
  type ActionCenterFollowUpSignal,
  type ActionCenterReviewMoment,
  type ActionCenterWorkspaceSummary,
} from '@/lib/action-center-shared-core'
import type { MemberRole } from '@/lib/types'

export type ExitCarrierStatus = 'active'
export type ExitCarrierRouteScope = 'exit_only'
export type ExitCarrierOwnerModel = 'explicit_named_owner'
export type ExitCarrierReviewDiscipline = 'follow_up_review_required'
export type ExitTriageStatus = 'nieuw' | 'bevestigd' | 'geparkeerd' | 'uitgevoerd' | 'verworpen'
export type ExitDeliveryMode = 'baseline' | 'live'

const OPEN_TRIAGE_STATUSES = new Set<ExitTriageStatus>(['nieuw', 'bevestigd'])

export interface ExitActionCenterCarrier {
  key: 'exit'
  label: string
  status: ExitCarrierStatus
  workspaceKind: 'follow_through'
  routeScope: ExitCarrierRouteScope
  ownerModel: ExitCarrierOwnerModel
  reviewDiscipline: ExitCarrierReviewDiscipline
  canOpenOtherProductAdapters: false
}

export interface ExitDossierInput {
  id: string
  title: string
  triageStatus: ExitTriageStatus
  deliveryMode: ExitDeliveryMode | null
  managementOwnerLabel: string | null
  reviewOwnerLabel: string | null
  firstActionTaken: string | null
  reviewMoment: string | null
  managementActionOutcome: string | null
  nextRoute: string | null
  stopReason: string | null
}

export interface ExitActionCenterWorkspace {
  carrier: ExitActionCenterCarrier
  summary: ActionCenterWorkspaceSummary
  dossiers: ActionCenterDossier[]
  assignments: ActionCenterAssignment[]
  reviewMoments: ActionCenterReviewMoment[]
  followUpSignals: ActionCenterFollowUpSignal[]
  activeDeliveryModes: ExitDeliveryMode[]
}

export interface ExitActionCenterCandidateDossier {
  scanType: string | null
  routeInterest: string | null
  campaignId: string | null
}

const EXIT_ACTION_CENTER_CARRIER: ExitActionCenterCarrier = {
  key: 'exit',
  label: 'ExitScan-adapter',
  status: 'active',
  workspaceKind: 'follow_through',
  routeScope: 'exit_only',
  ownerModel: 'explicit_named_owner',
  reviewDiscipline: 'follow_up_review_required',
  canOpenOtherProductAdapters: false,
}

function buildActorId(prefix: string, value: string | null) {
  if (!value) return null
  const normalized = value.trim().toLowerCase().split(/\s+/).join('-')
  return normalized ? `${prefix}:${normalized}` : null
}

function isOpen(triageStatus: ExitTriageStatus) {
  return OPEN_TRIAGE_STATUSES.has(triageStatus)
}

function isExitDeliveryMode(value: ExitDeliveryMode | null): value is ExitDeliveryMode {
  return value === 'baseline' || value === 'live'
}

export function getExitActionCenterCarrier() {
  return EXIT_ACTION_CENTER_CARRIER
}

export function isExitActionCenterCandidate(args: {
  dossier: ExitActionCenterCandidateDossier
  exitCampaignIds: ReadonlySet<string>
}) {
  const { dossier, exitCampaignIds } = args

  if (dossier.scanType === 'exit') {
    return true
  }

  if (dossier.campaignId) {
    return exitCampaignIds.has(dossier.campaignId)
  }

  return dossier.scanType === null && dossier.routeInterest === 'exitscan'
}

export function buildExitActionCenterWorkspace(args: {
  role: MemberRole | null | undefined
  dossiers: ExitDossierInput[]
}): ExitActionCenterWorkspace {
  const dossiers: ActionCenterDossier[] = []
  const assignments: ActionCenterAssignment[] = []
  const reviewMoments: ActionCenterReviewMoment[] = []
  const followUpSignals: ActionCenterFollowUpSignal[] = []
  const activeDeliveryModes = Array.from(
    new Set(args.dossiers.map((dossier) => dossier.deliveryMode).filter(isExitDeliveryMode)),
  ).sort((left, right) => left.localeCompare(right))

  for (const dossier of args.dossiers) {
    const permissionEnvelope = buildActionCenterPermissionEnvelope({ role: args.role })
    const open = isOpen(dossier.triageStatus)
    const hasFollowThroughDecision = Boolean(
      dossier.managementActionOutcome || dossier.nextRoute || dossier.stopReason,
    )
    const managementOwnerId = buildActorId('manager', dossier.managementOwnerLabel)
    const reviewOwnerId = buildActorId('review', dossier.reviewOwnerLabel)
    const ownerId = managementOwnerId ?? reviewOwnerId

    dossiers.push({
      id: dossier.id,
      title: dossier.title,
      status: open ? 'open' : 'closed',
      ownerId,
      permissionEnvelope,
    })

    const assignmentState =
      dossier.triageStatus === 'uitgevoerd'
        ? 'completed'
        : dossier.triageStatus === 'verworpen'
          ? 'cancelled'
          : dossier.triageStatus === 'geparkeerd'
            ? 'waiting'
            : dossier.firstActionTaken
              ? 'active'
              : 'blocked'
    const assignmentKind =
      hasFollowThroughDecision ? 'handoff' : dossier.firstActionTaken ? 'follow_up' : 'review_prep'

    assignments.push({
      id: `asg-${dossier.id}`,
      dossierId: dossier.id,
      title: dossier.firstActionTaken ?? 'Leg eerste ExitScan follow-up stap vast',
      state: assignmentState,
      kind: assignmentKind,
      ownerId,
    })

    const reviewState =
      dossier.triageStatus === 'uitgevoerd'
        ? 'completed'
        : dossier.triageStatus === 'geparkeerd' || dossier.triageStatus === 'verworpen'
          ? 'cancelled'
          : 'due'

    reviewMoments.push({
      id: `rev-${dossier.id}`,
      dossierId: dossier.id,
      scheduledFor: dossier.reviewMoment ?? 'Exit reviewmoment nog vastleggen',
      state: reviewState,
    })

    if (open && !ownerId) {
      followUpSignals.push({
        id: `sig-owner-${dossier.id}`,
        dossierId: dossier.id,
        kind: 'owner_missing',
        severity: 'warning',
        state: 'open',
      })
    }

    if (open && !dossier.firstActionTaken) {
      followUpSignals.push({
        id: `sig-step-${dossier.id}`,
        dossierId: dossier.id,
        kind: 'blocked_assignment',
        severity: 'critical',
        state: 'open',
      })
    }

    if (open) {
      followUpSignals.push({
        id: `sig-review-${dossier.id}`,
        dossierId: dossier.id,
        kind: 'review_due',
        severity: 'warning',
        state: 'open',
      })
    }

    if (open && !hasFollowThroughDecision) {
      followUpSignals.push({
        id: `sig-decision-${dossier.id}`,
        dossierId: dossier.id,
        kind: 'decision_due',
        severity: 'warning',
        state: 'open',
      })
    }
  }

  return {
    carrier: EXIT_ACTION_CENTER_CARRIER,
    summary: buildActionCenterWorkspaceSummary({
      dossiers,
      assignments,
      reviewMoments,
      followUpSignals,
    }),
    dossiers,
    assignments,
    reviewMoments,
    followUpSignals,
    activeDeliveryModes,
  }
}

File: backend/products/shared/action_center_exit.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from .action_center_core import (
    ActionAssignment,
    ActionDossier,
    ActionFollowUpSignal,
    ActionReviewMoment,
    ActionWorkspaceSummary,
    MemberRole,
    build_permission_envelope,
    summarize_workspace,
)


ExitCarrierStatus = Literal["active"]
ExitCarrierRouteScope = Literal["exit_only"]
ExitCarrierOwnerModel = Literal["explicit_named_owner"]
ExitCarrierReviewDiscipline = Literal["follow_up_review_required"]
ExitTriageStatus = Literal["nieuw", "bevestigd", "geparkeerd", "uitgevoerd", "verworpen"]
ExitDeliveryMode = Literal["baseline", "live"]

_OPEN_TRIAGE_STATUSES = {"nieuw", "bevestigd"}


@dataclass(frozen=True)
class ExitActionCenterCarrier:
    key: Literal["exit"]
    label: str
    status: ExitCarrierStatus
    workspace_kind: Literal["follow_through"]
    route_scope: ExitCarrierRouteScope
    owner_model: ExitCarrierOwnerModel
    review_discipline: ExitCarrierReviewDiscipline
    can_open_other_product_adapters: Literal[False]


@dataclass(frozen=True)
class ExitDossierInput:
    id: str
    title: str
    triage_status: ExitTriageStatus
    delivery_mode: ExitDeliveryMode | None
    management_owner_label: str | None
    review_owner_label: str | None
    first_action_taken: str | None
    review_moment: str | None
    management_action_outcome: str | None
    next_route: str | None
    stop_reason: str | None


@dataclass(frozen=True)
class ExitActionCenterWorkspace:
    carrier: ExitActionCenterCarrier
    summary: ActionWorkspaceSummary
    dossiers: list[ActionDossier]
    assignments: list[ActionAssignment]
    review_moments: list[ActionReviewMoment]
    follow_up_signals: list[ActionFollowUpSignal]
    active_delivery_modes: tuple[ExitDeliveryMode, ...]


_EXIT_ACTION_CENTER_CARRIER = ExitActionCenterCarrier(
    key="exit",
    label="ExitScan-adapter",
    status="active",
    workspace_kind="follow_through",
    route_scope="exit_only",
    owner_model="explicit_named_owner",
    review_discipline="follow_up_review_required",
    can_open_other_product_adapters=False,
)


def _build_actor_id(prefix: str, value: str | None) -> str | None:
    if not value:
        return None
    normalized = "-".join(value.strip().lower().split())
    return f"{prefix}:{normalized}" if normalized else None


def _is_open(triage_status: ExitTriageStatus) -> bool:
    return triage_status in _OPEN_TRIAGE_STATUSES


def _has_bounded_text(value: str | None) -> bool:
    return bool(value and value.strip())


def get_exit_action_center_carrier() -> ExitActionCenterCarrier:
    return _EXIT_ACTION_CENTER_CARRIER


def build_exit_action_center_workspace(
    *,
    role: MemberRole | None,
    dossiers: list[ExitDossierInput],
) -> ExitActionCenterWorkspace:
    shared_dossiers: list[ActionDossier] = []
    assignments: list[ActionAssignment] = []
    review_moments: list[ActionReviewMoment] = []
    follow_up_signals: list[ActionFollowUpSignal] = []
    active_delivery_modes = tuple(
        sorted({dossier.delivery_mode for dossier in dossiers if dossier.delivery_mode in {"baseline", "live"}})
    )

    for dossier in dossiers:
        permission_envelope = build_permission_envelope(role)
        is_open = _is_open(dossier.triage_status)
        has_follow_through_decision = bool(
            dossier.management_action_outcome or dossier.next_route or dossier.stop_reason
        )
        has_review_moment = _has_bounded_text(dossier.review_moment)
        management_owner_id = _build_actor_id("manager", dossier.management_owner_label)
        review_owner_id = _build_actor_id("review", dossier.review_owner_label)
        owner_id = management_owner_id or review_owner_id

        shared_dossiers.append(
            ActionDossier(
                id=dossier.id,
                title=dossier.title,
                status="open" if is_open else "closed",
                owner_id=owner_id,
                permission_envelope=permission_envelope,
            )
        )

        assignment_state = (
            "completed"
            if dossier.triage_status == "uitgevoerd"
            else "cancelled"
            if dossier.triage_status == "verworpen"
            else "waiting"
            if dossier.triage_status == "geparkeerd"
            else "active"
            if dossier.first_action_taken
            else "blocked"
        )
        assignment_kind = (
            "handoff"
            if has_follow_through_decision
            else "follow_up"
            if dossier.first_action_taken
            else "review_prep"
        )
        assignments.append(
            ActionAssignment(
                id=f"asg-{dossier.id}",
                dossier_id=dossier.id,
                title=dossier.first_action_taken or "Leg eerste ExitScan follow-up stap vast",
                state=assignment_state,
                kind=assignment_kind,
                owner_id=owner_id,
            )
        )

        review_state = (
            "completed"
            if dossier.triage_status == "uitgevoerd"
            else "cancelled"
            if dossier.triage_status in {"geparkeerd", "verworpen"}
            else "scheduled"
            if has_review_moment
            else "due"
        )
        review_moments.append(
            ActionReviewMoment(
                id=f"rev-{dossier.id}",
                dossier_id=dossier.id,
                scheduled_for=dossier.review_moment or "Exit reviewmoment nog vastleggen",
                state=review_state,
            )
        )

        if is_open and not owner_id:
            follow_up_signals.append(
                ActionFollowUpSignal(
                    id=f"sig-owner-{dossier.id}",
                    dossier_id=dossier.id,
                    kind="owner_missing",
                    severity="warning",
                    state="open",
                )
            )

        if is_open and not dossier.first_action_taken:
            follow_up_signals.append(
                ActionFollowUpSignal(
                    id=f"sig-step-{dossier.id}",
                    dossier_id=dossier.id,
                    kind="blocked_assignment",
                    severity="critical",
                    state="open",
                )
            )

        if is_open and not has_review_moment:
            follow_up_signals.append(
                ActionFollowUpSignal(
                    id=f"sig-review-{dossier.id}",
                    dossier_id=dossier.id,
                    kind="review_due",
                    severity="warning",
                    state="open",
                )
            )

        if is_open and not has_follow_through_decision:
            follow_up_signals.append(
                ActionFollowUpSignal(
                    id=f"sig-decision-{dossier.id}",
                    dossier_id=dossier.id,
                    kind="decision_due",
                    severity="warning",
                    state="open",
                )
            )

    summary = summarize_workspace(
        dossiers=shared_dossiers,
        assignments=assignments,
        review_moments=review_moments,
        follow_up_signals=follow_up_signals,
    )

    return ExitActionCenterWorkspace(
        carrier=_EXIT_ACTION_CENTER_CARRIER,
        summary=summary,
        dossiers=shared_dossiers,
        assignments=assignments,
        review_moments=review_moments,
        follow_up_signals=follow_up_signals,
        active_delivery_modes=active_delivery_modes,
    )
